rust/bridge/node/futures/src/future.rs

```diff
@@ -0,0 +1,205 @@
+//
+// Copyright 2020-2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use neon::prelude::*;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem;
+use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex, Weak};
+use std::task::{Poll, Waker};
+
+use crate::result::*;
+
+mod builder;
+pub use builder::JsFutureBuilder;
+
+/// A transformation to convert a scoped JavaScript result (fulfillment or rejection) to an unscoped Rust result.
+trait JsFutureCallback<T> =
+    for<'a> FnOnce(&mut FunctionContext<'a>, JsPromiseResult<'a>) -> T + 'static + Send;
+
+/// The possible states of a [JsFuture].
+enum JsFutureState<T> {
+    /// The future is waiting for resolution.
+    Waiting {
+        transform: Box<dyn JsFutureCallback<T>>,
```

Comment:
The transform has to be done while the JavaScript context is active, but when you `await` the JsFuture you'll be in plain Rust code somewhere, with no access to the context. So the value that's returned by `await` has to be produced at the point when the future is fulfilled so it can be picked up later.